### DISCUSSION TITLE
Installation problem with ffnet

### DISCUSSION THREAD

**Discussion by jvrod23** [spacepy/spacepy#591] [Support and Q&A] [answered]
Following a hardware upgrade, I'm having problems installing SpacePy.  I'm working in macOS Monterey on a MacBook Pro.  

The problem has to do with ffnet.  

When I run pip install spacepy

I get a very long error readout which I believe can be boiled down to:

**Building wheels for collected packages: spacepy, ffnet
  Building wheel for spacepy (pyproject.toml) ... done
  Created wheel for spacepy: filename=spacepy-0.2.3-cp39-cp39-macosx_12_0_x86_64.whl size=7405754 sha256=a07b81a95d6c9eb836f617686ea3a19d1918c61b1616e043719c098b56cfb82d
  Stored in directory: /Users/juan.rodriguez/Library/Caches/pip/wheels/38/19/0f/7fcdc756674e7c38bd50a92a69b4880cbc1dd055fb7fba3794
  Building wheel for ffnet (setup.py) ... error

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: legacy-install-failure

× Encountered error while trying to install package.
╰─> ffnet

note: This is an issue with the package mentioned above, not pip.**

Since I have the hard dependencies already installed, I thought I would try

pip install --no-deps spacepy

That seemed to work.  However, when I tried to import irbempy, I got the following error message from irbempy.py:

ImportError: cannot import name 'irbempylib' from partially initialized module 'spacepy.irbempy' (most likely due to a circular import) (/Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages/spacepy/irbempy/__init__.py)

Since I don't use lanlstar, I would appreciate advice on whether there is a workaround that will allow me to install SpacePy without ffnet, and still allow me to use irbempy.

**Comment by drsteve** [comment, marked as answer]
The "circular import" has cropped up for a few people, mostly on MacOS (I'm not sure why). I've had a hard time reproducing, but the relevant issues that may be of help are #535 and #585. The code shouldn't have any circular imports, so the install probably failed somewhere...  I know that we do need to make some installer changes to keep up with the latest changes in how Python expects packages to be installed and how they locate themselves. That's on our list, but I'm not sure whether it's the cause here.

One thing I would recommend checking is the contents of `/Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages/spacepy/irbempy/`. There should be a shared object library there (`.so` on linux, I think that's a `.dylib` on mac). Is that present? Missing? If present, what's it called?

My general recommendation would be to take a look at the two issues linked above, especially checking pathing (and looking in the installation output to see if it complains about pathing), and maybe install from source (if necessary doing the build separately before the install). Hopefully that'll get you back to a working install as fast as possible.

**Reply by jvrod23** [reply]
Sorry about the slow response.   This may not be a circular import issue.

/Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages/spacepy/irbempy does not contain a shared object library:

It contains two .py files and the __pycache__ directory:

__init__.py	__pycache__	irbempy.py

__pycache__ contains two .pyc files:  __init__.cpython-39.pyc	irbempy.cpython-39.pyc

This is the state following 

pip install --no-deps spacepy

The only .so file installed in the spacepy folder following this command is /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages/spacepy/libspacepy.so

So perhaps it is not a circular import issue, but rather that the irbempy .so file was not installed due to the --no-deps flag?

If I simply run pip install spacepy, nothing is installed, due presumably to the problem with the ffnet installation.

I've attached the complete verbose transcript from this installation attempt.  Apart from the failure of the wheel build for ffnet, there seem to be some C compilation errors.  I don't see any complaints about pathing.
[pip_install_spacepy_errors.txt](https://github.com/spacepy/spacepy/files/8522312/pip_install_spacepy_errors.txt)

**Reply by jtniehof** [reply]
The attached transcript fails at the attempt to build ffnet, which doesn't seem compatible with the `--no-deps` build. But the underlying issue appears to be the same as #583 and would result in a SpacePy build without irbem. For some reason f2py is looking for a library that's not on the path. This appears to be an f2py and/or xcode issue and @spacecataz noted he had to completely uninstall and reinstall xcode after an OS update.

**Reply by jvrod23** [reply]
Sorry, to be clear, the transcript was after pip install spacepy.  

pip install --no-deps spacepy returned no errors at all, but resulted in no irbempy.

I may need to get my administrator to uninstall and reinstall xcode on this NOAA computer.  This is a new computer with Monterey.  Although it is possible that xcode was installed before the upgrade to Monterey...it has been sitting on a shelf for a little while, waiting for a user.

I was able to install spacepy (and xcode) successfully on my personal laptop a few weeks ago after an OS upgrade to Monterey.

**Reply by jvrod23** [reply]
After Xcode uninstall and reinstall, the problem with ffnet install went away.  That was an important clue from one of the other issues.  That got rid of a lot of obfuscating issues.

Very benign feedback from pip install spacepy:

  Using cached spacepy-0.3.0.zip (19.8 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: numpy!=1.15.0,>=1.10 in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages (from spacepy) (1.22.3)
Requirement already satisfied: matplotlib>=1.5 in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages (from spacepy) (3.4.3)
Requirement already satisfied: h5py>=2.6 in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages (from spacepy) (3.6.0)
Requirement already satisfied: python-dateutil>=1.4 in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages (from spacepy) (2.8.2)
Requirement already satisfied: scipy>=0.11 in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages (from spacepy) (1.8.0)
Requirement already satisfied: kiwisolver>=1.0.1 in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5->spacepy) (1.3.1)
Requirement already satisfied: pyparsing>=2.2.1 in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5->spacepy) (3.0.4)
Requirement already satisfied: cycler>=0.10 in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5->spacepy) (0.10.0)
Requirement already satisfied: pillow>=6.2.0 in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages (from matplotlib>=1.5->spacepy) (8.4.0)
Requirement already satisfied: six>=1.5 in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages (from python-dateutil>=1.4->spacepy) (1.16.0)
Building wheels for collected packages: spacepy
  Building wheel for spacepy (pyproject.toml) ... done
  Created wheel for spacepy: filename=spacepy-0.3.0-cp39-cp39-macosx_12_0_x86_64.whl size=1185242 sha256=fc431a9243eca4db9cda652042f76ef638a69e7fbcc2555ab80668b833dbc3de
  Stored in directory: /Users/juan.rodriguez/Library/Caches/pip/wheels/58/87/83/adb83134b561567d6cffe35b176c191c41903fc597e3c0a894
Successfully built spacepy
Installing collected packages: spacepy
Successfully installed spacepy-0.3.0

However, the circular import error remains: 

ImportError: cannot import name 'irbempylib' from partially initialized module 'spacepy.irbempy' (most likely due to a circular import) (/Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages/spacepy/irbempy/__init__.py)

There is no shared object library in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages/spacepy/irbempy

Originally I thought this was due to using the --no-deps option but this is now demonstrated not to be the case.

Is install from source the best option now?

**Reply by jtniehof** [reply]
The ffnet issue probably went away because we removed ffnet as a dependency--it's become impossible to build with up-to-date libraries.

There are two possibilities here. One is the `-lSystem` issue, which your work with xcode should have addressed. The other is that we've seen weirdness where the `install` doesn't always call `build` properly. We haven't seen it with `pip`, but there's always a first time.

So I think a source install makes sense. Grab `spacepy-0.3.0.zip` from [our releases](https://github.com/spacepy/spacepy/releases) (ignore the one that says "source code", that's not actually a fully prepared distribution.) There are two ways you can go after uncompressing the source:

First, you can try a direct `python setup.py install`. That is _supposed to_ do the build first, but we've found that it doesn't always--so if you're willing to try for us, that might help us get another data point on the issue.

If that doesn't work, or you'd rather just jump to the thing with the highest likelihood of success, try:
```
python setup.py build
python setup.py install
```

It would be useful to capture the output so you can attach it if there are problems.

Thanks for your patience. There's some weirdness going on with Macs that we don't completely understand right now.

**Reply by jvrod23** [reply]
The direct 'python setup.py install' didn't work, nor did 'python setup.py build' (I didn't bother with the second install command).  But I think there are some valuable clues in the output.  Most of the output went to file but some went direct to the terminal, so I copied that to the end of the file, after 'FROM TERMINAL'.  To avoid confusion, I only attached the output from the direct 'python setup.py install' attempt.

In there, I see the following two major failures or errors:

(1) **irbemlib module failed. Try a different Fortran compiler? (--fcompiler)
irbemlib build produced no recognizable module. Try a different Fortran compiler? (--fcompiler)**

Which compiler do you recommend?  It looks like I have:
customize Gnu95FCompiler
Found executable /Users/juan.rodriguez/opt/anaconda3/bin/x86_64-apple-darwin13.4.0-gfortran
compiling Fortran sources
Fortran f77 compiler: /Users/juan.rodriguez/opt/anaconda3/bin/x86_64-apple-darwin13.4.0-gfortran -std=legacy -arch x86_64 -fPIC -march=nocona -mtune=core2 -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -O3 -funroll-loops
Fortran f90 compiler: /Users/juan.rodriguez/opt/anaconda3/bin/x86_64-apple-darwin13.4.0-gfortran -std=legacy -arch x86_64 -fPIC -march=nocona -mtune=core2 -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -O3 -funroll-loops
Fortran fix compiler: /Users/juan.rodriguez/opt/anaconda3/bin/x86_64-apple-darwin13.4.0-gfortran -Wall -g -ffixed-form -fno-second-underscore -std=legacy -arch x86_64 -fPIC -march=nocona -mtune=core2 -ftree-vectorize -fPIC -fstack-protector -O2 -pipe -O3 -funroll-loops

(2) **ld: library not found for -lSystem
collect2: error: ld returned 1 exit status**

You identified this as a potential issue.  Since the reinstalled xcode didn't fix this, how should I install it?

I'm very grateful for your help!

[python_setup.py_install_verbose.txt](https://github.com/spacepy/spacepy/files/8591556/python_setup.py_install_verbose.txt)
.

**Reply by drsteve** [reply]
For the `-lSystem` error, which I think is a configuration issue with xcode, please see https://github.com/spacepy/spacepy/discussions/583#discussioncomment-2335377

That should let the compilation proceed.

Also, note that the build command tries to build the package but won't put it anywhere useful, so the install command is required to complete installation. Technically just running install _should_ do both steps but we've had mixed results on Mac recently.

**Reply by jtniehof** [reply]
You're right that if the build failed, there's no point in running the install. @spacecataz , do you have more ideas for things to try on the `-lSystem` issue?

**Reply by spacecataz** [reply]
I have experience with this. It's almost certainly still an Xcode issue. My issue was with Xcode and gfortran, not macports. My solution required two essential steps: wipe BOTH Xcode AND Macports, then reinstall both in order. I also needed to install macports' gfortran from source instead of binary; this step was not necessary with more recent Macports releases.

Some time later, a graduate student ran into the SAME problem you're having now. She reinstalled Xcode to fix; see the detailed steps below.

**MY GUESS** is that you will need to reinstall any of the gfortran compilers (or C or anything else) that used the old Xcode install, as the system path is likely broken for those installs. This may mean uninstalling and reinstalling Conda.

It's not pretty. Pasting directly from my notes from some time ago to give you all of the details I had on the issue:

### MY ISSUE: XCODE, MACPORTS, AND GFORTRAN
Issue: Linking stage of compilation with GCC/gfortran (and others) fails because libraries aren't found.  The key error is library not found for -lsystem

See this discussion here: https://trac.macports.org/ticket/58895

This is an issue with Xcode.  Here are the *rough* steps I took to fix this, may need revision:
This is a scorched Earth approach; it may be possible to do this in fewer steps, but I had issues accomplishing this and only the scorched Earth approach worked.

- Uninstall gcc: `sudo port uninstall gcc9`
- Uninstall Xcode.
- Uninstall MacPorts. See this migration guide for saving your installed ports so that you can restore things quickly: https://trac.macports.org/wiki/Migration
- Reinstall Xcode. Be sure to follow the instructions on the Macports installation guide to ensure command line tools are ready, the license agreement has been handled, etc: https://guide.macports.org/#installing.xcode
- Reinstall MacPorts.  See the permission settings from the "Ls Operation not permitted" problem on another page of this notebook.
- Reinstall gcc from source (-v flag): `sudo port -v -s install gcc9`
- Follow the rest of the migration guide to reinstall old ports.
		
### STUDENT'S PROBLEM: Irbemlib install failure
Student ran into a similar problem where gfortran (via spacepy/irbemlib compilation) failed because it could not find system header files (specifically, stdio.h).  Reinstalling Xcode fixed.

Solution: https://donatstudios.com/MojaveMissingHeaderFiles

```
sudo rm -rf /Library/Developer/CommandLineTools
xcode-select --install
cd /Library/Developer/CommandLineTools/Packages/
open macOS_SDK_headers_for_macOS_10.14.pkg
```

**Reply by drsteve** [reply]
I would check out #583 first and just try adding the SDK location to `LIBRARY_PATH` and `LD_LIBRARY_PATH`. It's side-stepping the issue of xcode (and friends) needing full reinstalls, but it's definitely a faster path forward. Especially if you don't administer your own machine. That fixed the `-lSystem` issue for a student of mine.

**Reply by jtniehof** [reply]
I hate to do too much duck tape if xcode is in an unhappy place (since that may mask other issues), but if the SDK location is static, we could just shove that into the environment variables ourselves in `setup.py`.

**Reply by jvrod23** [reply]
I tried adding the following lines to .bashrc alone, and python setup.py install failed:

export LIBRARY_PATH="$LIBRARY_PATH:/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib"

export LD_LIBRARY_PATH="$LD_LIBRARY_PATH:/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/lib"

Then I added the lines to .bash_profile, and python setup.py install worked!  This produced irbempylib.cpython-39-darwin.so in /Users/juan.rodriguez/opt/anaconda3/lib/python3.9/site-packages/spacepy/irbempy/

I no longer get the irbempy 'circular import' complaint.  

I'm rerunning one of my analyses to make sure I can reproduce results I generated on my old laptop, before the replacement that led to this problem.  It has just completed analyzing one day of satellite data, so I'm optimistic that I can avoid the scorched Earth approach for now. But it's good to have it as a backup.  Thank you all!

**Reply by jtniehof** [reply]
I am closing this discussion as 0.4.0 (via #639) does this environment manipulation, and ffnet is no longer used. Further install issues are likely to be at least somewhat different.
